lib/floe/workflow.rb

```diff
@@ -5,40 +5,74 @@
 module Floe
   class Workflow
     class << self
-      def load(path_or_io, context = {}, credentials = {})
+      def load(path_or_io, context = nil, credentials = {})
         payload = path_or_io.respond_to?(:read) ? path_or_io.read : File.read(path_or_io)
         new(payload, context, credentials)
       end
     end
 
-    attr_reader :context, :credentials, :first_state, :payload, :states, :states_by_name, :start_at
+    attr_reader :context, :credentials, :payload, :states, :states_by_name, :current_state, :status
 
-    def initialize(payload, context = {}, credentials = {})
+    def initialize(payload, context = nil, credentials = {})
       payload     = JSON.parse(payload)     if payload.kind_of?(String)
       context     = JSON.parse(context)     if context.kind_of?(String)
       credentials = JSON.parse(credentials) if credentials.kind_of?(String)
+      context     = Context.new(context)    unless context.kind_of?(Context)
+
+      @payload     = payload
+      @context     = context || {"global" => {}}
+      @credentials = credentials
 
-      @payload        = payload
-      @context        = context
-      @credentials    = credentials
       @states         = payload["States"].to_a.map { |name, state| State.build!(self, name, state) }
-      @states_by_name = states.to_h { |state| [state.name, state] }
-      @start_at       = @payload["StartAt"]
-      @first_state    = @states_by_name[@start_at]
+      @states_by_name = @states.each_with_object({}) { |state, result| result[state.name] = state }
+      start_at        = @payload["StartAt"]
+
+      current_state_name = @context["State"]["Name"] || start_at
+      @current_state = @states_by_name[current_state_name]
+
+      @status = current_state_name == start_at ? "pending" : current_state.status
     rescue JSON::ParserError => err
       raise Floe::InvalidWorkflowError, err.message
     end
 
-    def run!
-      state = first_state
-      input = context.dup
+    def step
+      @status = "running" if @status == "pending"
+      @context["Execution"]["StartTime"] ||= Time.now.utc
+
+      input = @context["State"]["Output"] || @context["Execution"]["Input"].dup
+
+      tick = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+      next_state, output = current_state.run!(input)
+      tock = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+
+      @context["State"] = {
+        "EnteredTime"  => tick,
+        "FinishedTime" => tock,
+        "Duration"     => tock - tick,
+        "Output"       => output,
+        "Name"         => next_state&.name,
+        "Input"        => output
+      }
 
-      until state.nil?
-        state, output = state.run!(input)
-        input = output
+      @context["States"] << @context["State"]
```

Comment:
Think this needs to be a method on context

next_state = context.run_and_record(current_state)

Reply:
Hm I'm not sure about `run_and_record` going through context since it is the state that we're running and context just keeps some info about the run.
Maybe we could move some of this into the run method, but it would look basically the same just be in a different place and we have one or two states which have to entirely override the run method right now so it'd be duplicated code.

Reply:
Maybe the "return" that is set to `@context["State"]` could be returned by the `current_state.run!`?  I'll have to play with that a bit.